# Michael Theisen and Trevor Tomlin

# 1)

Using the following data “gene_expression_n438x978.txt” and 
“ADRs_HLGT_n438x232.txt” to answer this question.  
 
In class, we used forward stepwise logistic regression to build predictive models for side 
effect “heart failure” by considering only the first 20 genes (variables) in 
“gene_expression_n438x978.txt” and using all 438 drugs.   
 
Answer this question using all 438 drugs.

# 1a)
Start with the NULL model.  Consider the first 50 genes (columns 1:50 in 
“gene_expression_n438x978.txt”), use forward stepwise logistic regression to build 
predictive models for each of the 232 side effects in “ADRs_HLGT_n438x232.txt”.   
Which side effect can be predicted with the smallest number of errors? What is the 
corresponding model?
Which side effect can be predicted with the smallest AIC?  What is the corresponding 
model?

In [17]:
my.data <- read.table ("gene_expression_n438x978.txt", sep="\t", header=T)
dimnames(my.data)[[1]] <- my.data[,1]
my.data <- my.data[, -1]
my.data[1:5, 1:6]

PSME1        ATF1         RHEB         FOXO3      
ACETAZOLAMIDE       -0.015159099 -0.031470528 -0.004733488  0.02591061
IRBESARTAN          -0.026811981  0.012151979 -0.025550148 -0.02401181
IPRATROPIUM BROMIDE  0.001017958 -0.008650622 -0.018128698 -0.02079971
EFAVIRENZ           -0.004398264  0.055387992  0.004658520  0.01380732
THIAMINE             0.001838965 -0.018079188 -0.011855532 -0.03705033
                    RHOA          IL1B        
ACETAZOLAMIDE        0.0056296773 -0.007325439
IRBESARTAN          -0.0106717396  0.070951154
IPRATROPIUM BROMIDE -0.0002722781  0.025003360
EFAVIRENZ           -0.0340697348  0.024005885
THIAMINE            -0.0133954959 -0.001752450

In [18]:
adr.mat <- read.table ("ADRs_HLGT_n438x232.txt", sep="\t", header=T)
dimnames(adr.mat)[[1]] <- adr.mat [, 1]
adr.mat <- adr.mat[, -1]
adr.mat[1:5, 1:6]

Abdominal.hernias.and.other.abdominal.wall.conditions
ACETAZOLAMIDE       0                                                    
IRBESARTAN          0                                                    
IPRATROPIUM BROMIDE 0                                                    
EFAVIRENZ           0                                                    
THIAMINE            0                                                    
                    Abortions.and.stillbirth Acid.base.disorders
ACETAZOLAMIDE       0                        1                  
IRBESARTAN          0                        0                  
IPRATROPIUM BROMIDE 0                        0                  
EFAVIRENZ           0                        0                  
THIAMINE            0                        0                  
                    Administration.site.reactions Adrenal.gland.disorders
ACETAZOLAMIDE       1                             0                      
IRBESARTAN          0                             1                      
IPRATROPIUM BROMIDE 0                             1                      
EFAVIRENZ           0                             1                      
THIAMINE            0                             0                      
                    Age.related.factors
ACETAZOLAMIDE       0                  
IRBESARTAN          0                  
IPRATROPIUM BROMIDE 0                  
EFAVIRENZ           1                  
THIAMINE            0

In [19]:
options(warn=-1)

In [20]:
first.20.genes <- data.frame(my.data[ , 1:20]) # start with the subset of the first 20 genes.
first.20.genes[1:5,] # sanity check. prints out a table of the first 5 drugs vs the first 20 genes.

PSME1        ATF1         RHEB         FOXO3      
ACETAZOLAMIDE       -0.015159099 -0.031470528 -0.004733488  0.02591061
IRBESARTAN          -0.026811981  0.012151979 -0.025550148 -0.02401181
IPRATROPIUM BROMIDE  0.001017958 -0.008650622 -0.018128698 -0.02079971
EFAVIRENZ           -0.004398264  0.055387992  0.004658520  0.01380732
THIAMINE             0.001838965 -0.018079188 -0.011855532 -0.03705033
                    RHOA          IL1B         ASAH1        RALA        
ACETAZOLAMIDE        0.0056296773 -0.007325439  0.011877825  0.022163771
IRBESARTAN          -0.0106717396  0.070951154  0.003130203  0.010251465
IPRATROPIUM BROMIDE -0.0002722781  0.025003360 -0.011930075 -0.001534971
EFAVIRENZ           -0.0340697348  0.024005885  0.028270797  0.031744732
THIAMINE            -0.0133954959 -0.001752450 -0.005194435  0.023154467
                    ARHGEF12    SOX2         SERPINE1      HLA.DMA      
ACETAZOLAMIDE       0.015935689  0.001867957  0.0536455506  0.0161696678
IRBESARTAN          0.064158581  0.004275614 -0.0133962084 -0.0263783481
IPRATROPIUM BROMIDE 0.002130394  0.013207036 -0.0367360390 -0.0001869585
EFAVIRENZ           0.004021388 -0.007478017 -0.0051331342 -0.0224352845
THIAMINE            0.046257712  0.003875417  0.0008968498 -0.0235725707
                    EGF          APP          NOS3        CSNK1A1      
ACETAZOLAMIDE       -0.010473857  0.006264292  0.04489467 -0.0075977283
IRBESARTAN          -0.011073187 -0.007276688 -0.02599797 -0.0004622327
IPRATROPIUM BROMIDE  0.012464447 -0.010497377  0.00991886 -0.0110961892
EFAVIRENZ            0.002552100 -0.009437163  0.01499500 -0.0239943786
THIAMINE             0.009344182  0.004243532 -0.01223848 -0.0170609476
                    NFATC4       TBP          BRCA1        PSMD4       
ACETAZOLAMIDE        0.033024746 -0.009180880 -0.048011062 -0.020221419
IRBESARTAN           0.011080955 -0.018843837  0.013304076 -0.009440527
IPRATROPIUM BROMIDE  0.005871478  0.001191925 -0.003608489 -0.005478129
EFAVIRENZ            0.010403449  0.004904294  0.032500218 -0.009029595
THIAMINE            -0.016359751  0.013808377  0.004106289 -0.016174886

In [21]:
min.error <- 1
min.aic <- .Machine$integer.max
best.model.errors <- 0
best.model.aic <- 0
best.ind.error <- 0
best.ind.aic <- 0

for (x in 1:232) {
    null.model <- glm (adr.mat[,x] ~ 1, data=first.20.genes, family=binomial)
    first.20.genes.model <- glm (adr.mat[,x] ~ ., data=first.20.genes, family=binomial)
    model <- step (null.model, trace=0, scope=list(upper=first.20.genes.model), data=first.20.genes, direction="forward")
    pred <- predict(model, type="response")
    errors <- sum(adr.mat[,x] != round(pred))
    
    total <- length(adr.mat[,x])
    error.percent <- errors/total
    
    if (error.percent < min.error) {
        min.error <- error.percent
        best.model.error <- model
        best.ind.error <- x
    }
    
    if (model$aic < min.aic) {
        min.aic <- model$aic
        best.model.aic <- model
        best.ind.aic <- x
    }
 
}

In [22]:
print(min.error)
print(best.model.error)
pred <- predict(best.model.error, type="response")
print(table (adr.mat[,best.ind.error], round(pred)))

[1] 0.01598174

Call:  glm(formula = adr.mat[, x] ~ 1, family = binomial, data = first.20.genes)

Coefficients:
(Intercept)  
      -4.12  

Degrees of Freedom: 437 Total (i.e. Null);  437 Residual
Null Deviance:	    71.8 
Residual Deviance: 71.8 	AIC: 73.8
   
      0
  0 431
  1   7


In [23]:
print(colnames(adr.mat)[[best.ind.error]])

[1] "Gastrointestinal.tract.disorders.congenital"


**Gastrointestinal.tract.disorders.congenital** can calculated with an error rate of 0.01598174.\
The model is uses only the intercept.

In [24]:
print(min.aic)
print(best.model.aic)
pred <- predict(best.model.aic, type="response")
print(table (adr.mat[,best.ind.aic], round(pred)))

[1] 73.79585

Call:  glm(formula = adr.mat[, x] ~ 1, family = binomial, data = first.20.genes)

Coefficients:
(Intercept)  
      -4.12  

Degrees of Freedom: 437 Total (i.e. Null);  437 Residual
Null Deviance:	    71.8 
Residual Deviance: 71.8 	AIC: 73.8
   
      0
  0 431
  1   7


In [25]:
print(colnames(adr.mat)[[best.ind.aic]])

[1] "Gastrointestinal.tract.disorders.congenital"


**Gastrointestinal.tract.disorders.congenital** can calculated with an AIC of 73.79585.\
The model is uses only the intercept.

# 1b)

Repeat Question (2a) using backward elimination instead of forward 
stepwise.  Start with the FULL model.  Consider the first 50 genes (columns 1:50 in 
“gene_expression_n438x978.txt”), use backward elimination in logistic regression to 
build predictive models for each of the 232 side effects in 
“ADRs_HLGT_n438x232.txt”.   
Which side effect can be predicted with the smallest number of errors? What is the 
corresponding model? 
Which side effect can be predicted with the smallest AIC?  What is the corresponding 
model? 

In [26]:
min.error <- 1
min.aic <- .Machine$integer.max
best.model.errors <- 0
best.model.aic <- 0
best.ind.error <- 0
best.ind.aic <- 0

for (x in 1:232) {
    null.model <- glm (adr.mat[,x] ~ 1, data=first.20.genes, family=binomial)
    first.20.genes.model <- glm (adr.mat[,x] ~ ., data=first.20.genes, family=binomial)
    model <- step (first.20.genes.model, trace=0, data=first.20.genes, direction="backward")
    pred <- predict(model, type="response")
    errors <- sum(adr.mat[,x] != round(pred))
    
    total <- length(adr.mat[,x])
    error.percent <- errors/total
    
    if (error.percent < min.error) {
        min.error <- error.percent
        best.model.error <- model
        best.ind.error <- x
    }
    
    if (model$aic < min.aic) {
        min.aic <- model$aic
        best.model.aic <- model
        best.ind.aic <- x
    }
 
}

In [27]:
print(min.error)
print(best.model.error)
pred <- predict(best.model.error, type="response")
print(table (adr.mat[,best.ind.error], round(pred)))

[1] 0.01598174

Call:  glm(formula = adr.mat[, x] ~ FOXO3 + ASAH1 + BRCA1 + PSMD4, family = binomial, 
    data = first.20.genes)

Coefficients:
(Intercept)        FOXO3        ASAH1        BRCA1        PSMD4  
     -4.826       24.141       26.118       27.386      -21.644  

Degrees of Freedom: 437 Total (i.e. Null);  433 Residual
Null Deviance:	    71.8 
Residual Deviance: 63.76 	AIC: 73.76
   
      0
  0 431
  1   7


In [28]:
print(colnames(adr.mat)[[best.ind.error]])

[1] "Gastrointestinal.tract.disorders.congenital"


**Gastrointestinal.tract.disorders.congenital** can calculated with an error rate of 0.01598174.\
The model is uses the genes: FOXO3, ASAH1, BRCA1, PSMD4.

In [29]:
print(min.aic)
print(best.model.aic)
pred <- predict(best.model.aic, type="response")
print(table (adr.mat[,best.ind.aic], round(pred)))

[1] 73.75807

Call:  glm(formula = adr.mat[, x] ~ FOXO3 + ASAH1 + BRCA1 + PSMD4, family = binomial, 
    data = first.20.genes)

Coefficients:
(Intercept)        FOXO3        ASAH1        BRCA1        PSMD4  
     -4.826       24.141       26.118       27.386      -21.644  

Degrees of Freedom: 437 Total (i.e. Null);  433 Residual
Null Deviance:	    71.8 
Residual Deviance: 63.76 	AIC: 73.76
   
      0
  0 431
  1   7


In [30]:
print(colnames(adr.mat)[[best.ind.aic]])

[1] "Gastrointestinal.tract.disorders.congenital"


**Gastrointestinal.tract.disorders.congenital** can calculated with an AIC of 73.79585.\
The model is uses the genes: FOXO3, ASAH1, BRCA1, PSMD4.

# 2
In class, we discussed many techniques for classification and feature selection in the 
context of personalized medicine.  We illustrated how to apply these methods to the 
breast cancer data in class

# 2a)
Apply the t-test to the gene expression data and select the 50 most 
significant (i.e. lowest p-values) genes for each side effect. Repeat backward 
elimination using these 50 genes.

In [31]:
# For each side effect E:
#     For each gene G:
#         x = vector consisting of gene expression values of gene G with no ADR (i.e. response = 0) for E
#         y = vector consisting of gene expression values of gene G with ADR (ie. response = 1) for E
#         Apply the t-test. t.test (x,y)

best.aic <- .Machine$integer.max
best.e <- 0
best.model <- 0

for (e in 1:232) {
    gene.and.pvalues <- matrix(, nrow=0, ncol=2)
    
    for (g in 1:978) { 
        no.adr <- my.data[which(adr.mat[,e] == 0),g]
        adr <- my.data[which(adr.mat[,e] == 1),g]
        p.value <- t.test(no.adr, adr)$p.value
        data <- c(g,p.value)
        gene.and.pvalues <- rbind(gene.and.pvalues, data)
    
    } 
    gene.and.pvalues <- gene.and.pvalues[order(gene.and.pvalues[,2]),][1:20,]
    
    top.20.gene.matrix <- matrix(, nrow=438, ncol=0)
    
    for (i in gene.and.pvalues[,1]) {
        top.20.gene.matrix <- cbind(my.data[,2], top.20.gene.matrix)
    
    }
    
    df <- as.data.frame(top.20.gene.matrix)
    colnames(df) <- colnames(my.data)[gene.and.pvalues[,1]]

    # Backward step
    null.model <- glm (adr.mat[,e] ~ 1, data=df, family=binomial)
    top.genes.model <- glm (adr.mat[,e] ~ ., data=df, family=binomial)
    model <- step (top.genes.model, trace=0, data=df, direction="backward")
    
    if (model$aic < best.aic) {
        best.aic <- model$aic
        best.e <- e
        best.model <- model
    }

}

In [32]:
print(best.aic)
print(colnames(adr.mat)[[best.e]])
print(best.model)

[1] 73.79585
[1] "Gastrointestinal.tract.disorders.congenital"

Call:  glm(formula = adr.mat[, e] ~ 1, family = binomial, data = df)

Coefficients:
(Intercept)  
      -4.12  

Degrees of Freedom: 437 Total (i.e. Null);  437 Residual
Null Deviance:	    71.8 
Residual Deviance: 71.8 	AIC: 73.8


# 2b)
Compare your answer to Q1(b) and discuss any differences you observed. 

We see that that by using the t-test and selecting the best genes, we get the best model as seen above.  Gastrointestinal.tract.disorders.congenital has the best AIC as seen before. However, this time the NULL model has the same results as the models previously which were based on several genes.